lib/maths/CBoostedTreeFactory.cc

```diff
@@ -692,27 +716,96 @@ CBoostedTreeFactory::testLossLineSearch(core::CDataFrame& frame,
     // low chance of missing the best solution by doing so.
 
     using TMeanVarAccumulator = CBasicStatistics::SSampleMeanVar<double>::TAccumulator;
+    using TMinAccumulator = CBasicStatistics::SMin<double>::TAccumulator;
+
+    auto boptVector = [](double regularizer) {
+        return SConstant<CBayesianOptimisation::TVector>::get(1, regularizer);
+    };
+
+    TMinAccumulator minTestLoss;
+    TDoubleDoublePrVec testLosses;
+    testLosses.reserve(MAX_LINE_SEARCH_ITERATIONS);
+    // Ensure we choose one value based on expected improvement.
+    std::size_t minNumberTestLosses{5};
+
+    CBayesianOptimisation bopt{{{intervalLeftEnd, intervalRightEnd}}};
+    for (auto regularizer :
+         {intervalLeftEnd, (2.0 * intervalLeftEnd + intervalRightEnd) / 3.0,
+          (intervalLeftEnd + 2.0 * intervalRightEnd) / 3.0, intervalRightEnd}) {
+        if (applyRegularizer(*m_TreeImpl, regularizer) == false) {
+            m_TreeImpl->m_TrainingProgress.increment(
+                (MAX_LINE_SEARCH_ITERATIONS - testLosses.size()) * m_TreeImpl->m_MaximumNumberTrees);
+            break;
+        }
 
-    CLeastSquaresOnlineRegression<2, double> leastSquaresQuadraticTestLoss;
-    TDoubleVec testLosses;
-    testLosses.reserve(LINE_SEARCH_ITERATIONS);
-
-    for (std::size_t i = 0; i < LINE_SEARCH_ITERATIONS; ++i) {
-        if (applyRegularizerStep(*m_TreeImpl, stepSize, i) == false) {
+        CBoostedTreeImpl::TNodeVecVec forest;
+        double testLoss;
+        std::tie(forest, testLoss) = m_TreeImpl->trainForest(
+            frame, m_TreeImpl->m_TrainingRowMasks[0], m_TreeImpl->m_TestingRowMasks[0],
+            m_TreeImpl->m_TrainingProgress, m_RecordMemoryUsage);
+        bopt.add(boptVector(regularizer), testLoss, 0.0);
+        minTestLoss.add(testLoss);
+        testLosses.emplace_back(regularizer, testLoss);
+    }
+    while (testLosses.size() > 0 && testLosses.size() < MAX_LINE_SEARCH_ITERATIONS) {
```

Comment:
Yes. This can happen in the loop which searches for the best downsample factor, i.e. `applyRegularizer` can return false for all values in the loop
```
for (auto regularizer :
         {intervalLeftEnd, (2.0 * intervalLeftEnd + intervalRightEnd) / 3.0,
          (intervalLeftEnd + 2.0 * intervalRightEnd) / 3.0, intervalRightEnd}) {
  ...
```
In that case, we'll just use downsample factor of 1 and this actually SIGSEGVs when trying to compute maximum expected loss without this condition.